internal/ai/conversation.go

```diff
@@ -157,6 +164,120 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+	}
+
+	// Get token count from most recent response for logging
+	var totalTokens int64
+	if lastMsg := cc.Messages[len(cc.Messages)-1]; lastMsg.Response != nil {
+		totalTokens = lastMsg.Response.Usage.InputTokens + lastMsg.Response.Usage.CacheReadInputTokens
+	}
+
+	log.Printf("Conversation has %d messages and %d total tokens (input+cache read), summarizing...", 
+		len(cc.Messages), totalTokens)
+
+	// Preserve the first message (initial repository and task content)
+	numFirstMessagesToPreserve := 1
+	
+	// Ensure we have something to summarize
+	if len(cc.Messages) <= numFirstMessagesToPreserve {
+		return nil
+	}
+
+	// Generate summary using AI (this will add a summary request/response to the conversation)
+	summary, err := cc.generateConversationSummary(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to generate conversation summary: %w", err)
+	}
+
+	// Create a summary message to replace the middle portion
+	summaryMessage := conversationTurn{
+		UserMessage: anthropic.NewUserMessage(anthropic.NewTextBlock(summary)),
+		// No response - this is just context for future messages
+	}
+
+	// Reconstruct the conversation: preserved first messages + summary
+	newMessages := []conversationTurn{}
+	
+	// Add preserved first messages
+	newMessages = append(newMessages, cc.Messages[:numFirstMessagesToPreserve]...)
+	
+	// Add summary message
+	newMessages = append(newMessages, summaryMessage)
+
+	// Update the conversation
+	originalMessageCount := len(cc.Messages)
+	cc.Messages = newMessages
+
+	log.Printf("Conversation summarized: %d messages -> %d messages", 
+		originalMessageCount, len(cc.Messages))
+
+	return nil
+}
+
+// generateConversationSummary creates a summary of the conversation using AI
+func (cc *Conversation) generateConversationSummary(ctx context.Context) (string, error) {
+	// Build a summary request that will be sent as part of the current conversation
+	var summaryPrompt strings.Builder
+	summaryPrompt.WriteString("Please summarize all of the work you have done so far. Focus on:\n")
+	summaryPrompt.WriteString("1. Key decisions and changes made\n")
+	summaryPrompt.WriteString("2. Current state of the codebase\n")
+	summaryPrompt.WriteString("3. Any important context for continuing the work\n")
+	summaryPrompt.WriteString("4. Tools used and their outcomes\n\n")
+	summaryPrompt.WriteString("Please provide a comprehensive but concise summary that captures all important information needed to continue working effectively. ")
+	summaryPrompt.WriteString("There's no need to include the system prompt or initial repository information in your summary - focus on the actual work and changes made during our conversation.")
+
+	// Temporarily disable token limit to avoid recursive summarization during summary generation
+	originalTokenLimit := cc.tokenLimit
+	cc.tokenLimit = 0
+
+	// Send summary request as part of the original conversation
+	response, err := cc.SendMessage(ctx, anthropic.NewTextBlock(summaryPrompt.String()))
```

Comment:
We should send this without the cache control param, since we're throwing away the conversation history anyway